### PR TITLE
Ch808 migrate code in preparation to use clue

### PR DESCRIPTION
**Pull request by jluhrs**
This PR migrates the base Seqexec code to use Java 11 and cats-effect 3, in preparation to use Clue. The main goal was to have it compiling and running. Therefore the code may not be the prettiest. Improvements will be left for a future PR.

### CODE REVIEW

**Comment by cquiroz** [comment]
Quick comment, you could remove the packtracker action, we are moving away from it

File: build.sbt [hunk 1]
```diff
@@ -347,25 +351,8 @@ lazy val acm = project
...
-        if (out.exists && out.listFiles.nonEmpty) out.listFiles.map(_.lastModified).min
-        else Int.MaxValue
-      if (mod > cur) {
-        out.mkdirs
-        val err = cmd.run(ProcessLogger(log.info(_), log.error(_))).exitValue
-        if (err != 0) sys.error("xjc failed")
-      }
-      out.listFiles.toSeq
-    }.taskValue
+    xjcCommandLine ++= Seq("-p", "edu.gemini.epics.acm.generated"),
```
Comment (cquiroz) [inline build.sbt:354]: Neat

File: modules/engine/src/main/scala/observe/engine/Engine.scala [hunk 2]
```diff
@@ -460,13 +460,15 @@ class Engine[F[_]: MonadError[*[_], Throwable]: Logger, S, U](stateL: Engine.Sta
     initialState: S,
     f:            (A, S) => F[(S, B, Option[Stream[F, A]])]
   )(implicit ev:  Concurrent[F]): Stream[F, B] =
-    Stream.eval(fs2.concurrent.Queue.unbounded[F, Stream[F, A]]).flatMap { q =>
-      Stream.eval_(q.enqueue1(input)) ++
-        q.dequeue.parJoinUnbounded
+    Stream.eval(cats.effect.std.Queue.unbounded[F, Stream[F, A]]).flatMap { q =>
+      Stream.exec(q.offer(input)) ++
```
Comment (cquiroz) [inline modules/engine/src/main/scala/observe/engine/Engine.scala:464]: Hadn't seen `exec` before, what does it do?

File: modules/engine/src/test/scala/observe/engine/StepSpec.scala [hunk 3]
```diff
@@ -429,30 +425,31 @@ class StepSpec extends CatsSuite {
...
               .takeThrough(a => !isFinished(a._2.sequences(seqId).status))
               .compile
               .toVector
           )
       }
-      .unsafeRunSync()
 
-    val actionsCompleted = qss.map(_._1).collect { case SystemUpdate(x: Completed[_], _) => x }
-    assert(actionsCompleted.length == 4)
+    qss.map { x =>
```
Comment (cquiroz) [inline modules/engine/src/test/scala/observe/engine/StepSpec.scala:439]: I think these are indepedent IO asserts right? If so you need to chain them and return them all at the end or onle the last one will be executed
Like `assert(actionsCompleted.length == 4) *> assert(executionsCompleted...)`
Also you may want to use `assertEquals`

File: modules/engine/src/test/scala/observe/engine/StepSpec.scala [hunk 4]
```diff
@@ -245,21 +236,23 @@ class StepSpec extends CatsSuite {
...
 
-    inside(m.compile.last.unsafeRunSync()) { case Some(Sequence.State.Zipper(zipper, status, _)) =>
-      inside(zipper.focus.toStep) { case Step(_, _, _, _, List(ex1, ex2)) =>
-        assert(
-          Execution(ex1.toList).results.length == 3 && Execution(ex2.toList).actions.length == 1
-        )
+    m.compile.last.map {
+      inside(_) { case Some(Sequence.State.Zipper(zipper, status, _)) =>
+        inside(zipper.focus.toStep) { case Step(_, _, _, _, List(ex1, ex2)) =>
+          assert(
```
Comment (cquiroz) [inline modules/engine/src/test/scala/observe/engine/StepSpec.scala:250]: I think here only the last assert is verified

File: modules/engine/src/test/scala/observe/engine/StepSpec.scala [hunk 5]
```diff
@@ -185,7 +179,7 @@ class StepSpec extends CatsSuite {
     case _                       => false
   }
 
-  def runToCompletion(s0: TestState): Option[TestState] =
+  def runToCompletion(s0: TestState): IO[Option[TestState]] =
```
Comment (cquiroz) [inline modules/engine/src/test/scala/observe/engine/StepSpec.scala:182]: This name may not be appropriate anymore as we are not really running the IO as before

File: modules/giapi/src/test/scala/giapi/client/GiapiStatusSpec.scala [hunk 6]
```diff
@@ -134,8 +127,7 @@ final class GiapiStatusSpec extends CatsSuite {
         GmpStatus.closeGmpStatus(g) >> c.get[Int](intItemName)
       }
       .attempt
-      .unsafeRunSync() should matchPattern { case Left(GiapiException(_)) =>
-    }
+      .map(matchPattern { case Left(GiapiException(_)) => })
```
Comment (cquiroz) [inline modules/giapi/src/test/scala/giapi/client/GiapiStatusSpec.scala:130]: I don't think munit includes this. Is this still on scalatest?

File: modules/server/src/main/scala/observe/server/ODBSequencesLoader.scala [hunk 7]
```diff
@@ -20,7 +19,7 @@ import SeqEvent._
 import ObserveFailure.ObserveException
 import ObserveFailure.UnrecognizedInstrument
 
-final class ODBSequencesLoader[F[_]: ApplicativeError[*[_], Throwable]](
+final class ODBSequencesLoader[F[_]: Async](
```
Comment (cquiroz) [inline modules/server/src/main/scala/observe/server/ODBSequencesLoader.scala:22]: Weird that you need this change, What does need `Async`?

File: build.sbt [hunk 8]
```diff
@@ -276,7 +278,8 @@ lazy val observe_server     = project
         TestLibs.value,
         PPrint.value
       ) ++ MUnit.value ++ Http4s ++ Http4sClient ++ PureConfig ++ SeqexecOdb ++ Monocle.value ++ WDBAClient ++
-        Circe.value
+        Circe.value,
+    headerSources / excludeFilter := HiddenFileFilter || (file("modules/server") / "src/main/scala/pureconfig/module/http4s/package.scala").getName
```
Comment (rpiaggio) [inline build.sbt:282]: This is because we are using a copy of the pureconfig-http4s bridge?

File: modules/server/src/main/scala/observe/server/Systems.scala [hunk 9]
```diff
@@ -384,7 +383,7 @@ object Systems {
   }
 
   private def decodeTops(s: String): Map[String, String] =
-    s.split("=|,")
+    s.split("[=,]")
```
Comment (cquiroz) [inline modules/server/src/main/scala/observe/server/Systems.scala:386]: Is this related to xjc change?

File: modules/server/src/test/scala/observe/server/InstrumentControllerSimSpec.scala [hunk 10]
```diff
@@ -51,120 +37,100 @@ class InstrumentControllerSimSpec extends CatsSuite {
...
-      r <- f.join
-    } yield r).unsafeRunSync()
+      _ <- IO.sleep(tick) // give it enough time for at least one tick
+      r <- f.joinWithNever
+    } yield assertEquals(r, ObserveCommandResult.Paused)
 
     // FIXME The simulator should return Stopped
-    // assert(result === ObserveCommandResult.Stopped)
-    assert(result === ObserveCommandResult.Paused)
+    // assertEquals(r, ObserveCommandResult.Stopped)
```
Comment (cquiroz) [inline modules/server/src/test/scala/observe/server/InstrumentControllerSimSpec.scala:108]: Commented code

File: modules/server/src/test/scala/observe/server/tcs/GuideConfigDbSuite.scala [hunk 11]
```diff
@@ -151,12 +151,9 @@ final class GuideConfigDbSuite extends munit.FunSuite {
   }
```
Comment (cquiroz) [inline modules/server/src/test/scala/observe/server/tcs/GuideConfigDbSuite.scala:151]: Weird that you can still have `shouldBe` under `munit`

File: modules/web/server/src/main/scala/observe/web/server/config/package.scala [hunk 12]
```diff
@@ -36,49 +34,48 @@ package config {
  */
 package object config {
 
-  implicit val siteReader = ConfigReader.fromCursor[Site] { cf =>
+  implicit val siteReader: AnyRef with ConfigReader[Site] = ConfigReader.fromCursor[Site] { cf =>
```
Comment (cquiroz) [inline modules/web/server/src/main/scala/observe/web/server/config/package.scala:37]: Why the `AnyRef`?

File: modules/web/server/src/main/scala/observe/web/server/logging/AppenderForClients.scala [hunk 13]
```diff
@@ -20,7 +20,8 @@ import observe.model.events._
  *
  * This is out of the scala/http4s loop
  */
-class AppenderForClients(out: Topic[IO, ObserveEvent]) extends AppenderBase[ILoggingEvent] {
+class AppenderForClients(out: Topic[IO, ObserveEvent])(implicit ioruntime: IORuntime)
```
Comment (cquiroz) [inline modules/web/server/src/main/scala/observe/web/server/logging/AppenderForClients.scala:23]: Should we use `Dispatcher` here instead?

File: modules/web/server/src/test/scala/observe/web/server/http4s/ObserveUIApiRoutesSpec.scala [hunk 14]
```diff
@@ -5,36 +5,35 @@ package observe.web.server.http4s
...
     // see https://github.com/http4s/http4s/issues/234
-    val r = (for {
+    for {
       s <- uiRoutes
       r <- Nested(
              s.apply(Request(method = Method.GET, uri = uri("/observe/login"))).value
            ).map(_.status).value
-    } yield r).unsafeRunSync()
-    assert(r === Some(Status.NotFound))
+    } yield assert(r === Some(Status.NotFound))
```
Comment (cquiroz) [inline modules/web/server/src/test/scala/observe/web/server/http4s/ObserveUIApiRoutesSpec.scala:32]: `assertEquals`

File: build.sbt (hunk 8 above)
Comment (jluhrs) [reply build.sbt:282]: Yes, and I did not want to stamp the AURA license on "borrowed" code.

**Review by cquiroz — CHANGES_REQUESTED** [review]
Great work, This is a big update with many moving parts interacting. I added some comments mostly about tests
Should we copy these changes to the seqexec?

File: modules/server/src/test/scala/observe/server/InstrumentControllerSimSpec.scala (hunk 10 above)
Comment (jluhrs) [reply modules/server/src/test/scala/observe/server/InstrumentControllerSimSpec.scala:108]: But, that commented code is relevant to the FIXME

File: modules/server/src/test/scala/observe/server/tcs/GuideConfigDbSuite.scala (hunk 11 above)
Comment (jluhrs) [reply modules/server/src/test/scala/observe/server/tcs/GuideConfigDbSuite.scala:151]: I missed that one, because it compiled nevertheless (because of an exiting import)

File: modules/web/server/src/main/scala/observe/web/server/config/package.scala (hunk 12 above)
Comment (jluhrs) [reply modules/web/server/src/main/scala/observe/web/server/config/package.scala:37]: I don't remember writing that. I think it was put when I used an automatic fix for untyped public vars

File: modules/engine/src/main/scala/observe/engine/Engine.scala (hunk 2 above)
Comment (jluhrs) [reply modules/engine/src/main/scala/observe/engine/Engine.scala:464]: It executes an F[Unit] for its side effect only. In this case, to add an element in the `Queue`

File: modules/engine/src/test/scala/observe/engine/StepSpec.scala (hunk 3 above)
Comment (jluhrs) [reply modules/engine/src/test/scala/observe/engine/StepSpec.scala:439]: @cquiroz I may be misunderstanding something, or using the wrong asserts, but those come from munit. They all produce `Unit`, and handle test failures through side effects.

File: modules/giapi/src/test/scala/giapi/client/GiapiStatusSpec.scala (hunk 6 above)
Comment (jluhrs) [reply modules/giapi/src/test/scala/giapi/client/GiapiStatusSpec.scala:130]: Yes, the source code imports just that one matcher.

File: modules/server/src/main/scala/observe/server/Systems.scala (hunk 9 above)
Comment (jluhrs) [reply modules/server/src/main/scala/observe/server/Systems.scala:386]: No, it was just a suggestion from the IDE. The expressions are equivalent.

File: modules/server/src/main/scala/observe/server/ODBSequencesLoader.scala (hunk 7 above)
Comment (jluhrs) [reply modules/server/src/main/scala/observe/server/ODBSequencesLoader.scala:22]: There are some calls inside the code that require `Temporal`, and others that require `Sync`. `Async` groups both of them.

File: modules/web/server/src/main/scala/observe/web/server/logging/AppenderForClients.scala (hunk 13 above)
Comment (jluhrs) [reply modules/web/server/src/main/scala/observe/web/server/logging/AppenderForClients.scala:23]: Yes, we should. I'll change it.

**Review by cquiroz — APPROVED** [review]
Thanks LGTM

**Comment by swalker2m** [comment]
Will these updates make it into the current seqexec in the interim?

**Comment by cquiroz** [comment]
I think we should to ease maintenance of the seqexec in the interim. @jluhrs We also sue CE in the client, was that ported too?
